### PR TITLE
interleave

### PR DESCRIPTION
**Pull request by strake**

### CODE REVIEW

**Comment by treeowl** [comment]
`noDuplicate#` didn't have a token-polymorphic type until GHC 8.2, when I generalized it to fix a bug in lazy `ST`. You'll have to `unsafeCoerce` your way to glory in earlier versions.

**Comment by treeowl** [comment]
I added a commit to fix that. Feel free to use it or not.

**Comment by strake** [comment]
LGTM

**Comment by treeowl** [comment]
The build is broken on pre-AMP versions. We need to import `Control.Applicative` for those, or switch from `*>` to `>>`.

**Comment by strake** [comment]
@RyanGlScott Done, PTAL

**Comment by strake** [comment]
@RyanGlScott Done, PTAL

File: changelog.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## Changes in version 0.?.?.?
```
Comment (RyanGlScott) [inline changelog.md]: There's already a "Changes in version 0.?.?.?" section below. Put the mention of `unsafeInterleave` in that section instead.

Comment (treeowl) [reply changelog.md]: The likely reason for confusion is that we already have two, one for a bug-fix release and one for the next major or minor release.

Comment (RyanGlScott) [reply changelog.md]: Oh dear, I didn't notice that! In that case, I'm not exactly sure myself where this mention of `unsafeInterleave` should go...

**Comment by treeowl** [comment]
It should go in the top-most mystery section. It's a new feature, so if there's a bug-fix-only release it won't include this.

**Comment by strake** [comment]
Done

**Comment by RyanGlScott** [comment]
LGTM. Thanks!

**Pull request merged**